map.go

```diff
@@ -0,0 +1,233 @@
+type MutexMap struct {
+	options        *mutexMapOptions
+	db             *sql.DB
+	lock           sync.Mutex
+	countedMutexes map[string]*countedMutex
+}
+
+// countedMutex objects are held in MutexMap. They keep hold counters and the underlying Mutex.
+type countedMutex struct {
+	mutex       mutexOperations
+	timeoutLock chan struct{}
+	waiters     int32
+	references  int32
+}
+
+// lock locks the underlying Mutex after first acquiring the private lock.
+func (cm *countedMutex) lock(
+	ctx context.Context,
+	allocator func() (mutexOperations, error),
+	maxWaiters int32,
+	name string,
+) (context.Context, error) {
+	// First, try non-blocking lock. We only become a waiter if we can't acquire lock in non-blocking fashion.
+	select {
+	case cm.timeoutLock <- struct{}{}:
+		// Lock acquired. Will unlock when unlock is called.
+	default:
+		// Unable to acquire local lock without waiting. We are now a waiter.
+		// Check limits before waiting in a blocking fashion.
+		currentWaiters := atomic.AddInt32(&cm.waiters, 1)
+		defer atomic.AddInt32(&cm.waiters, -1)
+		if maxWaiters >= 0 && currentWaiters > maxWaiters {
+			return nil, dbmerr.NewMaxWaitersExceededError(int(maxWaiters), name)
+		}
+		select {
+		case cm.timeoutLock <- struct{}{}:
+			// Lock acquired. Will unlock when unlock is called.
+		case <-ctx.Done():
+			// timeout
+			return nil, ctx.Err()
+		}
+	}
+
+	if cm.mutex == nil {
+		m, err := allocator()
+		if err != nil {
+			// unlock our local lock
+			<-cm.timeoutLock
+			return nil, err
+		}
+		cm.mutex = m
+	}
+	lockCtx, err := cm.mutex.Lock(ctx)
+	if err != nil {
+		// unlock our local lock
+		<-cm.timeoutLock
+	}
+	return lockCtx, err
+}
+
+func (cm *countedMutex) unlock(ctx context.Context) error {
+	err := cm.mutex.Unlock(ctx)
+	// Unlock local lock even if the underlying implementation fails during unlock. The assumption
+	// is that underlying implementation should be able to handle a subsequent Lock call and provide
+	// mutual exclusion.
+	<-cm.timeoutLock
+	return err
+}
+
+// NewMutexMap allocates a new MutexMap. The passed db will be used for all database operations. options can be
+// used to customize behaviour.
+func NewMutexMap(db *sql.DB, options ...MutexMapOption) *MutexMap {
+	mmo := &mutexMapOptions{
+		maxLocalWaiters: -1,
+	}
+	for _, o := range options {
+		o(mmo)
+	}
+	if mmo.maxLocalWaiters < -1 {
+		mmo.maxLocalWaiters = -1
+	}
+	if mmo.allocator == nil {
+		mmo.allocator = dbMutexAllocator
+	}
+	return &MutexMap{
+		options:        mmo,
+		db:             db,
+		countedMutexes: make(map[string]*countedMutex),
+	}
+}
+
+// Lock locks named Mutex. If not already available for the given name, an underlying
+// Mutex will be allocated and kept for later use. In order to lock with a timeout pass ctx that
+// has a deadline.  The returned context can be used to detect if the lock is lost.
+func (mm *MutexMap) Lock(ctx context.Context, name string) (context.Context, error) {
```

Comment:
You think namespace may potentialy be a requirement?  Currently not an issue but for monolith schemas, it may be.  Or I guess leave the naming scheme for the api caller to deal with.

Reply:
I think that we can handle it with `WithMutexOptions`. Essentially, during initialization (call to `NewMutexMap`) the caller can use a different table name.  So each set of locks can have their own namespace. https://github.com/dynata/go-dbmutex/blob/map/map.go#L166